### PR TITLE
feat(main): experimental.tsconfig

### PR DESCRIPTION
**Pull request by florian-lefebvre**
## Changes

- Supersedes #11722
- Adds a new `experimental.typescript` option. That will generate a `./astro/tsconfig.json`
- Generated tsconfig contains a reference to `./astro/types.d.ts`, removing the need for `env.d.ts`
- If the root tsconfig does not extend the generated tsconfig, an error is thrown with a diff
- Due to TS limitations, we need to own `include` and `exclude`. If they're present in the root tsconfig, a warning will be shown
- `outDir` is now excluded by default but user can opt-out with an option
- Plan for Astro 5 is to unflag this

## Testing

Added test

## Docs

Changeset, error, config reference

### CODE REVIEW

File: .changeset/empty-fans-reflect.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+export default defineConfig({
+    experimental: {
+        typescript: {
++            include: ["foo"],
++            exclude: ["bar"]
+        }
+    }
+})
+```
```
Comment (ArmandPhilippot) [inline .changeset/empty-fans-reflect.md:49]: There is a small typo on `error` (`eror` > `error`) that needs to be fixed. Other than that, I'm not Sarah 😁 but I think this should a little reworded to be more direct and more precise (the two diff blocks):
````suggestion
Enabling this option will require a few updates to your root `tsconfig.json`:

1. `extends` must point to `./.astro/tsconfig.json` otherwise an error will be thrown with a diff to help you.
2. If set, move `include` and `exclude` to `experimental.typescript.include` and `experimental.typescript.exclude` respectively in your Astro config file.

For example, in `tsconfig.json` you'll need to make the following changes:

```diff
{
+    "extends": "./.astro/tsconfig.json"
-    "include": ["foo"],
-    "exclude": ["bar"]
}
```

Then you'll need to reflect them in `astro.config.mjs`:

```diff
import { defineConfig } from "astro/config"

export default defineConfig({
    experimental: {
        typescript: {
+            include: ["foo"],
+            exclude: ["bar"]
        }
    }
})
```
````

I'm not sure about the ``in `astro.config.mjs`:`` since we can have different extensions (e.g. `.ts`) but this will help the user to understand where these diff take place.

But again, I'm not Sarah or Chris... so maybe you'll have better suggestions. 😄

File: packages/astro/src/core/config/config.ts [hunk 2]
```diff
@@ -165,5 +166,29 @@ export async function resolveConfig(
 		throw e;
 	}
 
+	// TODO: update examples default tsconfigs
+	if (astroConfig.experimental.typescript) {
+		const rootTsConfigUrl = new URL('./tsconfig.json', astroConfig.root);
+		if (!fsMod.existsSync(rootTsConfigUrl)) {
```
Comment (Princesseuh) [inline packages/astro/src/core/config/config.ts:172]: There's a wonky thing people do sometimes where they'll intentionally not have a tsconfig in every package of their monorepo, to share a TypeScript project between them (Starlight does this, for instance)

It's super edge casey and not worth considering, imo, but still could be a problem when this is unflagged

File: packages/astro/src/core/config/settings.ts [hunk 3]
```diff
@@ -134,3 +145,64 @@ export async function createSettings(config: AstroConfig, cwd?: string): Promise
...
+	try {
+		const { typescript } = settings.config.experimental;
+		if (!typescript) {
+			return;
+		}
+
+		let newConfig = { ...rawConfig };
+
+		if (!rawConfig.extends) {
+			newConfig.extends = ['astro/tsconfigs/base', GENERATED_TSCONFIG_PATH];
```
Comment (Princesseuh) [inline packages/astro/src/core/config/settings.ts:159]: It's okay to not use our presets, I'd suggest adding the preset only if there's no tsconfig at all

File: .changeset/empty-fans-reflect.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
+---
+'astro': minor
+---
+
+Adds a new `experimental.typescript` configuration option
```
Comment (sarah11918) [inline .changeset/empty-fans-reflect.md:5]: Just noting before I get into a full review that this name is pretty generic, and can make it look like *TypeScript* is experimental in Astro. :joy:

Usually these names are specific to what they do. If this is to provide more fine-grained TS config options, maybe even `experimental.tsconfig` which makes it super clear that this is a new way to "handle" these options!

Comment (florian-lefebvre) [reply .changeset/empty-fans-reflect.md:5]: Good point! This is inspired by nuxt which have a `typescript` option but it makes more stuff so yeah I think `experimental.tsconfig` makes sense. I'll update

Comment (florian-lefebvre) [reply .changeset/empty-fans-reflect.md]: Done!

File: packages/astro/src/@types/astro.ts [hunk 5]
```diff
@@ -2408,6 +2408,133 @@ export interface AstroUserConfig {
...
+		 * @docs
+		 * @name experimental.tsconfig
+		 * @type {object}
+		 * @default `undefined`
+		 * @version 4.15.0
+		 * @description
+		 *
+		 * Enables the generation of `.astro/tsconfig.json`. This allows to exclude the `outDir` by default to avoid `astro check` false positives.
+		 *
+		 * If you enable this option, you'll need to update your `tsconfig.json` with a new `extends` value:
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2422]:
```suggestion
		 * To enable this option in an existing project, you will need to update your existing `tsconfig.json` to recognize the new file, and also potentially your `astro.config.mjs` as some items have now moved to Astro configuration options.
		 *
		 * Your `tsconfig.json` file must now extend both your TypeScript template (`base`, `strict`, or `strictest`) and the new `/.astro/` file:
```

File: packages/astro/src/@types/astro.ts [hunk 6]
```diff
@@ -2408,6 +2408,133 @@ export interface AstroUserConfig {
...
+		 *
+		 * ```json title="tsconfig.json" del={2} ins={3}
+		 * {
+		 *   extends: 'astro/tsconfigs/base',
+		 * 	 extends: ['astro/tsconfigs/base', './.astro/tsconfig.json']
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2427]:
``````suggestion
		 * ```json title="tsconfig.json" ins="'./.astro/tsconfig.json'"
		 * {
		 * 	 extends: ['astro/tsconfigs/base', './.astro/tsconfig.json']
``````
This looks simpler to me? Just highlight the new entry? Or is it because of the array you wanted an entirely new line?

File: packages/astro/src/@types/astro.ts [hunk 7]
```diff
@@ -2408,6 +2408,133 @@ export interface AstroUserConfig {
...
+		 * }
+		 * ```
+		 *
+		 * If you have `include` or `exclude` inside your root `tsconfig.json`, you'll need to move them to your Astro config:
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2431]:
```suggestion
		 * If you have `include` or `exclude` inside your root `tsconfig.json`, move them to your Astro config:
```

File: packages/astro/src/@types/astro.ts [hunk 8]
```diff
@@ -2408,6 +2408,133 @@ export interface AstroUserConfig {
...
+		 *
+		 * ```json title="tsconfig.json" del={2-3}
+		 * {
+		 *   include: ['foo'],
+		 *   exclude: ['bar']
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2436]: Can we use something real here instead of foo and bar? Seeing a real example helps people know: Are these file paths? If so, absolute/relative? Are these expressions?

File: packages/astro/src/@types/astro.ts [hunk 9]
```diff
@@ -2408,6 +2408,133 @@ export interface AstroUserConfig {
...
+		tsconfig?: {
+			/**
+			 * @docs
+			 * @name experimental.tsconfig.include
+			 * @type {Array<string>}
+			 * @default `undefined`
+			 * @version 4.15.0
+			 * @description
+			 *
+			 * If you have `include` your root `tsconfig.json`, you'll need to move it to your Astro config:
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2462]: This and the others, are missing a **definition**. This is an instruction, but not a definition for reference. Can you give me a one-line reference that explains what this property *is* first? (And the others)

File: packages/astro/src/@types/astro.ts [hunk 10]
```diff
@@ -2408,6 +2408,133 @@ export interface AstroUserConfig {
...
+		 *    }
+		 *   }
+		 * })
+		 * ```
+		 */
+		tsconfig?: {
+			/**
+			 * @docs
+			 * @name experimental.tsconfig.include
+			 * @type {Array<string>}
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2457]:
```suggestion
			 * kind: h4
			 * @type {Array<string>}
```
These are all missing the heading level (I know, I still have to add those docs to Astro Docs Docs!)

Headings are rendered as `h2`, regular entries are rendered as `h3`, to have sub properties, you'll need to add the heading level manually.

File: packages/astro/src/core/errors/errors-data.ts [hunk 11]
```diff
@@ -1302,6 +1302,18 @@ export const UnknownFilesystemError = {
...
 
+/**
+ * @docs
+ * @description
+ * Your `tsconfig.json` `extends` property is invalid
+ */
+export const TSConfigInvalidExtends = {
+	name: 'TSConfigInvalidExtends',
+	title: 'Your `tsconfig.json` `extends` property is invalid',
+	message: 'Your `tsconfig.json` `extends` property is invalid',
```
Comment (sarah11918) [inline packages/astro/src/core/errors/errors-data.ts:1313]: I don't think title and message should be the same. Use Erika's error guide! https://github.com/withastro/astro/blob/main/packages/astro/src/core/errors/README.md

File: packages/astro/src/core/errors/errors-data.ts [hunk 12]
```diff
@@ -1302,6 +1302,18 @@ export const UnknownFilesystemError = {
...
+	hint: 'Check out what change you need to make in the terminal above.',
```
Comment (sarah11918) [inline packages/astro/src/core/errors/errors-data.ts:1314]:
```suggestion
	hint: 'Check the terminal output for more detail.',
```
Something like this maybe? If you can guide them more closely to exactly what they should see in the terminal (e.g. "exactly what went wrong", exactly what caused the error) then I'd do that.

If this error is only reached when there's e.g. a typo in one of the two things you have to extend (the astro TS template, and the new file), then the hint could be something like "Check the [`experimental.tsconfig` documentation]() for the proper `extends` configuration"

**Review by sarah11918 — COMMENTED** [review]
OK, thanks for your patience @florian-lefebvre ! I think just a few things to clean up here, in the API reference, then that will help us make sure the changeset is consistent, uses the same wording/phrases etc!

File: packages/astro/src/@types/astro.ts (hunk 8 above)
Comment (florian-lefebvre) [reply packages/astro/src/@types/astro.ts:2439]:
```suggestion
		 *   include: ['src/**/*'],
		 *   exclude: ['node_modules']
```

I linked the official docs on Discord but I'm going to link them here as well: https://www.typescriptlang.org/tsconfig/#include. So those are relative to the project root

File: packages/astro/src/@types/astro.ts (hunk 9 above)
Comment (florian-lefebvre) [reply packages/astro/src/@types/astro.ts:2465]: Should we reuse the official tsconfig docs + link them? (also for exclude)

File: packages/astro/src/core/errors/errors-data.ts (hunk 12 above)
Comment (florian-lefebvre) [reply packages/astro/src/core/errors/errors-data.ts:1314]: It's showing a diff to help the user, your suggestion lgtm

File: packages/astro/src/core/config/settings.ts [hunk 13]
```diff
@@ -134,3 +145,63 @@ export async function createSettings(config: AstroConfig, cwd?: string): Promise
...
+				'types',
+				`Your root "tsconfig.json" has an "include" field. This will break types, please move it to your Astro config experimental.tsconfig.include option`,
+			);
+		}
+		if (rawConfig.exclude) {
+			logger.warn(
+				'types',
+				`Your root "tsconfig.json" has an "exclude" field. This will break types, please move it to your Astro config experimental.tsconfig.exclude option`,
+			);
+		}
```
Comment (florian-lefebvre) [inline packages/astro/src/core/config/settings.ts:182]: @sarah11918 2 warnings here as well!

File: packages/astro/src/@types/astro.ts [hunk 14]
```diff
@@ -2408,6 +2408,137 @@ export interface AstroUserConfig {
...
+		 * @docs
+		 * @name experimental.tsconfig
+		 * @type {object}
+		 * @default `undefined`
+		 * @version 4.15.0
+		 * @description
+		 *
+		 * Enables a streamlined way of configuring TypeScript in your project for better DX when setting up Astro for Intellisense, type-checking, and more.
+		 *
+		 * This option removes the need for a separate `src/env.d.ts` file to provide Vite's client types to your project. Instead, `/.astro/tsconfig.json` will be autogenerated with linked types for you. Notably, this fixes an existing issue with `astro check` false positives by excluding `outDir` by default.
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2422]:
```suggestion
		 * This option removes the need for a separate `src/env.d.ts` file to provide Vite's client types to your project. Instead, `/.astro/tsconfig.json` will be autogenerated with linked types for you. Notably, this fixes an existing issue with `astro check` false positives by excluding `outDir` from your checked files by default.
```

File: packages/astro/src/@types/astro.ts (hunk 8 above)
Comment (sarah11918) [reply packages/astro/src/@types/astro.ts:2439]: Love this change, perfect!

File: packages/astro/src/@types/astro.ts [hunk 15]
```diff
@@ -2408,6 +2408,137 @@ export interface AstroUserConfig {
...
+		 * ```
+		 *
+		 * ```js title="astro.config.*" ins={6-7}
+		 * import { defineConfig } from 'astro/config'
+		 *
+		 * export default defineConfig({
+		 *   experimental: {
+		 *    tsconfig: {
+		 *      include: ['foo'],
+		 *      exclude: ['bar']
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2450]:
```suggestion
		 *      include: ['src/**/*'],
		 *      exclude: ['node_modules']
```

File: packages/astro/src/@types/astro.ts [hunk 16]
```diff
@@ -2408,6 +2408,137 @@ export interface AstroUserConfig {
...
+			 * import { defineConfig } from 'astro/config'
+			 *
+			 * export default defineConfig({
+			 *   experimental: {
+			 *    tsconfig: {
+			 *      include: ['foo']
+			 *    }
+			 *   }
+			 * })
+			 * ```
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2484]:
``````suggestion
			 *
			 * Specifies an array of filenames or patterns to include in the program. These filenames are resolved relative to the directory containing the `tsconfig.json` file.
			 *
			 * ```js title="astro.config.*" {6}
			 * import { defineConfig } from 'astro/config'
			 *
			 * export default defineConfig({
			 *   experimental: {
			 *    tsconfig: {
			 *      include: ['src/**/*']
			 *    }
			 *   }
			 * })
			 * ```
``````
This is the API reference for the entry. We don't have to show upgrade guidance here, only what it is. We've shown how to convert above!

File: packages/astro/src/@types/astro.ts [hunk 17]
```diff
@@ -2408,6 +2408,137 @@ export interface AstroUserConfig {
...
+			 * import { defineConfig } from 'astro/config'
+			 *
+			 * export default defineConfig({
+			 *   experimental: {
+			 *    tsconfig: {
+			 *      exclude: ['bar']
+			 *    }
+			 *   }
+			 * })
+			 * ```
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2514]:
``````suggestion
			 *
			 * Specifies an array of filenames or patterns that should be skipped when resolving [`include`](#include). It only excludes items that would otherwise be checked by your `include` settings, and cannot be used to prevent other files from being checked. See [the official TypeScript documentation](https://www.typescriptlang.org/tsconfig/#exclude) for more details.
			 *
			 * ```js title="astro.config.*" {6}
			 * import { defineConfig } from 'astro/config'
			 *
			 * export default defineConfig({
			 *   experimental: {
			 *    tsconfig: {
			 *      include: ['src/**/*'],
			 *      exclude:['src/exclude/**/*']
			 *    }
			 *   }
			 * })
			 * ```
``````
From the definition at the link it looks like whatever is excluded must be something that was going to otherwise be included. Can you update the example to make sense, given that?

File: packages/astro/src/@types/astro.ts [hunk 18]
```diff
@@ -2408,6 +2408,137 @@ export interface AstroUserConfig {
...
+
+			/**
+			 * @docs
+			 * @name experimental.tsconfig.excludeOutDir
+			 * @type {boolean}
+			 * @default `true`
+			 * @version 4.15.0
+			 * @description
+			 *
+			 * By default, `outDir` will be added to excluded tsconfig files. You can opt-out of this behavior by setting this option to `false`:
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2526]:
```suggestion
			 * Whether or not to skip checking built files located in your `outDir` folder. By default, `outDir` will be excluded during operations such as `astro check`. You can opt-out of this behavior by explicitly setting this option to `false`:
```

File: packages/astro/src/core/config/settings.ts [hunk 19]
```diff
@@ -134,3 +145,63 @@ export async function createSettings(config: AstroConfig, cwd?: string): Promise
...
+			throw createTsconfigError(logger, rawConfig, newConfig);
+		} else if (!rawConfig.extends.includes(GENERATED_TSCONFIG_PATH)) {
+			newConfig.extends = [...rawConfig.extends, GENERATED_TSCONFIG_PATH];
+			throw createTsconfigError(logger, rawConfig, newConfig);
+		}
+
+		if (rawConfig.include) {
+			logger.warn(
+				'types',
+				`Your root "tsconfig.json" has an "include" field. This will break types, please move it to your Astro config experimental.tsconfig.include option`,
```
Comment (sarah11918) [inline packages/astro/src/core/config/settings.ts:174]:
```suggestion
				`Your root "tsconfig.json" has an "include" field. Please move it to your Astro config experimental.tsconfig.include option to avoid breaking types.`,
```

File: packages/astro/src/core/config/settings.ts [hunk 20]
```diff
@@ -134,3 +145,63 @@ export async function createSettings(config: AstroConfig, cwd?: string): Promise
...
+			);
+		}
+		if (rawConfig.exclude) {
+			logger.warn(
+				'types',
+				`Your root "tsconfig.json" has an "exclude" field. This will break types, please move it to your Astro config experimental.tsconfig.exclude option`,
```
Comment (sarah11918) [inline packages/astro/src/core/config/settings.ts:180]:
```suggestion
				`Your root "tsconfig.json" has an "exclude" field. Please move it to your Astro config experimental.tsconfig.exclude option to avoid breaking types.`,
```

File: .changeset/empty-fans-reflect.md [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+```
+
+By default, `outDir` will now be excluded to avoid a long standing issue when using `astro check` on an already built project. If you want to opt-out of this behavior, you can set `experimental.typescript.excludeOutDir` to `false`:
+
+```diff
+import { defineConfig } from "astro/config"
+
+export default defineConfig({
+    experimental: {
+        typescript: {
```
Comment (bluwy) [inline .changeset/empty-fans-reflect.md:63]:
```suggestion
        tsconfig: {
```

File: .changeset/empty-fans-reflect.md [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+    experimental: {
+        tsconfig: {
++            include: ["foo"],
++            exclude: ["bar"]
+        }
+    }
+})
+```
+
+By default, `outDir` will now be excluded to avoid a long standing issue when using `astro check` on an already built project. If you want to opt-out of this behavior, you can set `experimental.typescript.excludeOutDir` to `false`:
```
Comment (bluwy) [inline .changeset/empty-fans-reflect.md:56]: Is there a discussion about needing `excludeOutDir`? I think we should be strict here and not introduce that option. It's unlikely that a user wants to not exclude the out dir I think

File: packages/astro/src/@types/astro.ts (hunk 8 above)
Comment (florian-lefebvre) [reply packages/astro/src/@types/astro.ts:2439]: I think it still makes sense yes, if you wanted not to check root ts files like the astro config

File: packages/astro/src/@types/astro.ts [hunk 23]
```diff
@@ -2408,6 +2408,131 @@ export interface AstroUserConfig {
...
+		 * @docs
+		 * @name experimental.tsconfig
+		 * @type {object}
+		 * @default `undefined`
+		 * @version 4.15.0
+		 * @description
+		 *
+		 * Enables a streamlined way of configuring TypeScript in your project. Enjoy an improved DX when setting up Astro for Intellisense, type-checking, and more.
+		 *
+		 * This option removes the need for a separate `src/env.d.ts` file to provide Vite's client types to your project. Instead, `/.astro/tsconfig.json` will be autogenerated with linked types for you. Notably, this fixes an existing issue with `astro check` false positives by excluding `outDir` from your checked files by default.
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2422]:
```suggestion
		 * This option removes the need for a separate `src/env.d.ts` file to provide Vite's client types to your project. Instead, `/.astro/tsconfig.json` will be autogenerated with linked types for you. Notably, this fixes an existing issue with `astro check` false positives by excluding your build from your checked files by default.
```
 Trying to make this as little jargony as I can (since we do explain the config item later). Can we say this?

File: packages/astro/src/core/config/settings.ts [hunk 24]
```diff
@@ -134,3 +145,63 @@ export async function createSettings(config: AstroConfig, cwd?: string): Promise
 
 	return settings;
 }
+
+function validateTsconfig(settings: AstroSettings, logger: Logger, rawConfig: TSConfig) {
```
Comment (bluwy) [inline packages/astro/src/core/config/settings.ts:149]: Personally I'm not sure about validating the tsconfig ourselves. Tsconfig can come in a lot of kinds and configurations, and it may not always be possible to enforce `extends` to exist, nor 100% sure that the `include` and `exclude` are not set. For example, here's Vite's setup: https://github.com/vitejs/vite/blob/cfb621e7a5a3e24d710a9af156e6855e73caf891/packages/create-vite/template-react-ts/tsconfig.json#L1-L8

The feature is experimental so I don't mind if we want to try this, but I think it would be better to focus on improving the docs on this instead.

File: packages/astro/src/@types/astro.ts [hunk 25]
```diff
@@ -2408,6 +2408,126 @@ export interface AstroUserConfig {
...
+		 *
+		 * export default defineConfig({
+		 *   experimental: {
+		 *    tsconfig: {
+		 *      include: ['src/**/*'],
+		 *      exclude: ['node_modules']
+		 *    }
+		 *   }
+		 * })
+		 * ```
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2454]:
```suggestion
		 * ```
		 * 
		 * For a complete overview, and to give feedback on this experimental API,
     * see the [Feature RFC](https://github.com/withastro/roadmap/blob/actions/proposals/00xx-feature.md).
```
Is there a link it makes sense to put here?

File: packages/astro/src/core/config/config.ts (hunk 2 above)
Comment (bluwy) [reply packages/astro/src/core/config/config.ts:172]: I'm not sure it's correct to expect the tsconfig to be beside the root here either, it can be outside of the root and upwards as typescript can be configured that way. In which case we shouldn't automatically create the `tsconfig.json` as it'll override the config set in that different place out of root.

File: packages/astro/src/@types/astro.ts (hunk 25 above)
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2454]: Just to clarify, we haven't said what to do if you're starting from a new project! Does this affect *any* new projects yet, because they'd automatically be configured the "normal" way using `create astro` right? 

I'm trying to figure out if it's possible for this to even happen, and I don't think so, so I *think* it's safe to frame this as only converting existing projects?

File: .changeset/empty-fans-reflect.md (hunk 21 above)
Comment (sarah11918) [reply .changeset/empty-fans-reflect.md:63]: Thanks Bjorn! I haven't even *got* to the changeset until the API docs are finalized :smile:

File: .changeset/empty-fans-reflect.md (hunk 22 above)
Comment (florian-lefebvre) [reply .changeset/empty-fans-reflect.md:56]: I don't remember exactly where/when it occurred but @Princesseuh told me some people actually want to check their build folder for some reason, although it's rare

File: packages/astro/src/core/errors/errors-data.ts (hunk 11 above)
Comment (sarah11918) [reply packages/astro/src/core/errors/errors-data.ts:1313]: Yes, message should be an extended version of the title. "What happened, why etc."

File: packages/astro/src/core/config/settings.ts (hunk 24 above)
Comment (florian-lefebvre) [reply packages/astro/src/core/config/settings.ts:149]: The goal here was to help the user. Like if it comes out of experimental in 5.0, users may wonder why their types stopped working all of sudden. Explaining it in the terminal as early as possible sounds like the perfect time to me but happy to revisit.

Regarding vite's setup, I'm not sure what impact references have

File: packages/astro/src/@types/astro.ts (hunk 25 above)
Comment (florian-lefebvre) [reply packages/astro/src/@types/astro.ts:2454]: Unfortunately it's not part of any RFC so nothing to link

Comment (florian-lefebvre) [reply packages/astro/src/@types/astro.ts:2454]: if you're starting a new project, you have no include nor exclude and experimental.typescript is not enabled so no action is required on the user's end I believe. So yeah I think it's safe to frame it as is

File: packages/astro/src/core/config/config.ts (hunk 2 above)
Comment (florian-lefebvre) [reply packages/astro/src/core/config/config.ts:172]: I think that idea was from @matthewp. Happy to remove this behavior

File: packages/astro/src/core/errors/errors-data.ts (hunk 11 above)
Comment (florian-lefebvre) [reply packages/astro/src/core/errors/errors-data.ts:1313]:
```suggestion
	title: 'Your `tsconfig.json` `extends` property is invalid',
	message: '`experimental.tsconfig` requires specific `extends` values inside `tsconfig.json`. Check out what change you need to make in the terminal above.',
```

File: packages/astro/src/core/config/settings.ts (hunk 24 above)
Comment (bluwy) [reply packages/astro/src/core/config/settings.ts:149]: My concern is that tsconfig can be wildly different based on how the user wants to configure it. e.g. maybe they want to have `tsconfig.json -> tsconfig.base.json -> .astro/tsconfig.json`, or if they use [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) and the extends is somewhere else like `tsconfig.app.json`. 

So I don't think it's in our territory to validate and ensure they have this specific shape. Also, will this be a breaking change in Astro 5 when enabled by default? Will `env.d.ts` not work anymore?

File: packages/astro/src/core/config/settings.ts [hunk 26]
```diff
@@ -134,3 +145,63 @@ export async function createSettings(config: AstroConfig, cwd?: string): Promise
...
+	try {
+		if (!settings.config.experimental.tsconfig) {
+			return;
+		}
+
+		let newConfig = { ...rawConfig };
+
+		if (!rawConfig.extends) {
+			newConfig.extends = [GENERATED_TSCONFIG_PATH];
```
Comment (bluwy) [inline packages/astro/src/core/config/settings.ts:158]: The path to the generate tsconfig path isn't always `./.astro/tsconfig.json` if the user has a different root. If we want this, we'd need to take in account of that. Related to https://github.com/withastro/astro/pull/11832#discussion_r1732952227

File: .changeset/empty-fans-reflect.md [hunk 27]
```diff
@@ -0,0 +1,70 @@
+---
+'astro': minor
+---
+
+Adds a new `experimental.tsconfig` configuration option
```
Comment (sarah11918) [inline .changeset/empty-fans-reflect.md:5]: Note from me that Florian will now grab content from the finalized API documentation to refine the changeset! :raised_hands: 

I'm out until I'm summoned again to look at this!

File: packages/astro/src/core/config/settings.ts (hunk 24 above)
Comment (florian-lefebvre) [reply packages/astro/src/core/config/settings.ts:149]: Ah I see what you mean. As for `env.d.ts`, the plan is to stop writing it in favor of just the tsconfig

Comment (ematipico) [reply packages/astro/src/core/config/settings.ts:149]: I am with @bluwy here. This could get hairy if you step into a corporate project where there are many shapes fo configuration files. Let's remember that there's also `jsconfig.json` too.

So yeah... I suggest to not do it

File: packages/astro/src/core/sync/write-files.ts [hunk 28]
```diff
@@ -74,3 +80,36 @@ async function setUpEnvTs(settings: AstroSettings, fs: typeof fsMod, logger: Log
 		logger.info('types', `Added ${bold(envTsPathRelativetoRoot)} type declarations`);
 	}
 }
+
+async function setupTsconfig(settings: AstroSettings, fs: typeof fsMod, logger: Logger) {
+	const tsconfig = settings.config.experimental.tsconfig!;
```
Comment (ematipico) [inline packages/astro/src/core/sync/write-files.ts:85]: Please add a comment explaining why the bang operator is safe here. I presume because the function is always called when the experimental flag is on.

File: packages/astro/src/core/config/config.ts (hunk 2 above)
Comment (ematipico) [reply packages/astro/src/core/config/config.ts:172]: Let's leave the generation of `tsconfig.json` to the user and `tsc`. They should handle that. (AKA separation of concerns)

**Review by ematipico — APPROVED** [review]
Overall the feature is fine, however I would not trigger (for now) the logic that creates/updates/deletes any `tsconfig.json`

File: .changeset/empty-fans-reflect.md (hunk 22 above)
Comment (Princesseuh) [reply .changeset/empty-fans-reflect.md:56]: Yea, some people once wanted it, I think it could be excluded and those people wouldn't necessarily lose sleep over it, though. We can reconsider if someone else ask

File: packages/astro/src/@types/astro.ts [hunk 29]
```diff
@@ -2408,6 +2408,126 @@ export interface AstroUserConfig {
...
+			 * export default defineConfig({
+			 *   experimental: {
+			 *    tsconfig: {
+			 *      excludeOutDir: false
+			 *    }
+			 *   }
+			 * })
+			 * ```
+			 */
+			excludeOutDir?: boolean;
```
Comment (florian-lefebvre) [inline packages/astro/src/@types/astro.ts:2529]: @Princesseuh fyi @bluwy asked if we really need this

File: packages/astro/src/@types/astro.ts [hunk 30]
```diff
@@ -2408,6 +2408,126 @@ export interface AstroUserConfig {
...
+
+			/**
+			 * @docs
+			 * @name experimental.tsconfig.excludeOutDir
+			 * @type {boolean}
+			 * @default `true`
+			 * @version 4.15.0
+			 * @description
+			 *
+			 * Whether or not to skip checking built files located in your `outDir` folder. By default, `outDir` will be excluded during operations such as `astro check`. You can opt-out of this behavior by explicitly setting this option to `false`:
```
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2515]:
```suggestion
			 * Whether or not to skip checking built files located in your `outDir` folder. By default, `outDir` will be excluded during operations such as `astro check`. You can opt out of this behavior by explicitly setting this option to `false`:
```

File: .changeset/empty-fans-reflect.md [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+```
+
+It will create references under the hood, making `env.d.ts` optional if you haven't made any change to it:
+
+```diff
+-/// <reference types="astro/client" />
+-/// <reference path="../.astro/types.d.ts" />
+```
+
+To learn more, check out the [documentation](https://docs.astro.build/en/reference/configuration-reference/#experimentaltypescript).
```
Comment (sarah11918) [inline .changeset/empty-fans-reflect.md:53]:
```suggestion
To learn more, check out the [`experimental.tsconfig` documentation](https://docs.astro.build/en/reference/configuration-reference/#experimentaltypescript).
```

File: .changeset/empty-fans-reflect.md [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+    experimental: {
++        tsconfig: {
++            include: ['src/**/*'],
++            exclude: ['node_modules']
++        }
+    }
+})
+```
+
+It will create references under the hood, making `env.d.ts` optional if you haven't made any change to it:
```
Comment (sarah11918) [inline .changeset/empty-fans-reflect.md:46]:
```suggestion
It will create references under the hood, making `env.d.ts` only necessary if you have added / for... :
```
Can you phrase it this way? What is this file still required for?

Comment (florian-lefebvre) [reply .changeset/empty-fans-reflect.md:46]: Updated, not 100% about what to put. @Princesseuh wdyt?

Comment (Princesseuh) [reply .changeset/empty-fans-reflect.md:46]: Sarah's suggestion is good for me!

Comment (sarah11918) [reply .changeset/empty-fans-reflect.md:46]: Good by me for docs! Sounds legit. :joy: Erika can say if it's the correct thing to say.

File: .changeset/empty-fans-reflect.md [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+    experimental: {
++        tsconfig: {
++            include: ['src/**/*'],
++            exclude: ['node_modules']
++        }
+    }
+})
+```
+
+It will create references under the hood, making `env.d.ts` only necessary if you have added custom type declarations:
```
Comment (sarah11918) [inline .changeset/empty-fans-reflect.md:46]:
```suggestion
It will create references under the hood, making `env.d.ts` necessary only if you have added custom type declarations:
```
nit!

**Review by sarah11918 — APPROVED** [review]
Approving for docs! Thank you @florian-lefebvre !

**Review by Princesseuh — CHANGES_REQUESTED** [review]
Still not sure on that one, with all the concerns, the value of it keeps diminishing to the point where there's.. well, no point.

**Comment by florian-lefebvre** [comment]
You will be missed, dear `experimental.tsconfig`
